src/main/java/ir/sahab/kafkaconsumer/SmartCommitKafkaConsumer.java

```
@@ -0,0 +1,408 @@
+            }
+        };
+        this.rebalanceListener = new ConsumerRebalanceListener() {
+            @Override
+            public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                if (!partitions.isEmpty()) {
+                    logger.warn("Kafka consumer previous assignment revoked: {}", partitions);
+                }
+            }
+
+            @Override
+            public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                logger.info("Kafka consumer partitions assigned: {}", partitions);
+                offsetTracker.reset();
+            }
+        };
+        this.unappliedAcks = new ArrayBlockingQueue<>(MAX_UNAPPLIED_ACKS);
+    }
+
+    /**
+     * Subscribes to the given topic by this consumer. Partitions will be assigned dynamically.
+     * Either this method or {@link #assign(String, List)} should be
+     * called just once after construction and before calling {@link #start()}.
+     */
+    public void subscribe(String topic) {
+        checkState(this.topic == null, "It is already subscribed/assigned. You should call either "
+                + "subscribe() or assign() just once.");
+        checkArgument(!Strings.isNullOrEmpty(topic));
+
+        kafkaConsumer.subscribe(Collections.singleton(topic), rebalanceListener);
+        this.topic = topic;
+    }
+
+    /**
+     * Manually assigns the given partition form the requested topic to this consumer.
+     * Either this method or {@link #subscribe(String)} should be
+     * called just once after construction and before calling {@link #start()}.
+     */
+    public void assign(String topic, List<Integer> partitions) {
+        checkState(this.topic == null, "It is already subscribed/assigned. You should call either "
+                + "subscribe() or assign() just once.");
+        checkArgument(!Strings.isNullOrEmpty(topic));
+        requireNonNull(partitions);
```

I think this is not needed, throwing NullPointerException when parameter is null is better than throwing IllegalArguementException as it's more relevant for the case. Optionally u can put @throws NullPointerException to javadoc to note that this exception is thrown if either of parameters are null.

---

requireNonNull(...) throws NullPointerException. If I had not this exception, I would throw NullPointerException anyway but this practice is used just to fail faster with a simpler call stack. 
But as I said in another comment, it is very obvious exception and it is not necessary to document it.

---

Yeah u r right, I thought that requireNonNull throws IllegalArgumentException, my mistake.